docs/codeql/codeql-language-guides/analyzing-data-flow-in-swift.rst

```diff
@@ -0,0 +1,291 @@
+      method.hasQualifiedName("String", "init(format:_:)") and
+      sinkExpr = call.getArgument(0).getExpr() and
+      DataFlow::localFlow(DataFlow::exprNode(sourceExpr), DataFlow::exprNode(sinkExpr))
+    select sourceExpr, sinkExpr
+
+We can vary the source, for example, making the source the parameter of a function rather than an expression. The following query finds where a parameter is used for the format:
+
+.. code-block:: ql
+
+    import swift
+    import codeql.swift.dataflow.DataFlow
+
+    from CallExpr call, Method method, ParamDecl sourceParam, Expr sinkExpr
+    where
+      call.getStaticTarget() = method and
+      method.hasQualifiedName("String", "init(format:_:)") and
+      sinkExpr = call.getArgument(0).getExpr() and
+      DataFlow::localFlow(DataFlow::parameterNode(sourceParam), DataFlow::exprNode(sinkExpr))
+    select sourceParam, sinkExpr
+
+The following example finds calls to ``String.init(format:_:)`` where the format string is not a hard-coded string literal:
+
+.. code-block:: ql
+
+    import swift
+    import codeql.swift.dataflow.DataFlow
+
+    from CallExpr call, Method method, DataFlow::Node sinkNode
+    where
+      call.getStaticTarget() = method and
+      method.hasQualifiedName("String", "init(format:_:)") and
+      sinkNode.asExpr() = call.getArgument(0).getExpr() and
+      not exists(StringLiteralExpr sourceLiteral |
+        DataFlow::localFlow(DataFlow::exprNode(sourceLiteral), sinkNode)
+      )
+    select call, "Format argument to " + method.getName() + " isn't hard-coded."
+
+Global data flow
+----------------
+
+Global data flow tracks data flow throughout the entire program, and is therefore more powerful than local data flow.
+However, global data flow is less precise than local data flow, and the analysis typically requires significantly more time and memory to perform.
+
+.. pull-quote:: Note
+
+   .. include:: ../reusables/path-problem.rst
+
+Using global data flow
+~~~~~~~~~~~~~~~~~~~~~~
+
+You can use the global data flow library by implementing the module ``DataFlow::ConfigSig``:
+
+.. code-block:: ql
+
+   import codeql.swift.dataflow.DataFlow
+
+   module MyDataFlowConfiguration implements DataFlow::ConfigSig {
+     predicate isSource(DataFlow::Node source) {
+       ...
+     }
+
+     predicate isSink(DataFlow::Node sink) {
+       ...
+     }
+   }
+
+   module MyDataFlow = DataFlow::Global<MyDataFlowConfiguration>;
+
+These predicates are defined in the configuration:
+
+-  ``isSource`` - defines where data may flow from.
+-  ``isSink`` - defines where data may flow to.
+-  ``isBarrier`` - optionally, restricts the data flow.
+-  ``isAdditionalFlowStep`` - optionally, adds additional flow steps.
+
+The last line (``module MyDataFlow = ...``) instantiates the parameterized module for data flow analysis by passing the configuration to the parameterized module. Data flow analysis can then be performed using ``MyDataFlow::flow(DataFlow::Node source, DataFlow::Node sink)``:
```

Comment:
I think we can simplify this to something like:

```suggestion
The last line (``module MyDataFlow = ...``) passes the configuration to the parameterized module to instantiate the parameterized module ready for data flow analysis. Data flow analysis can then be performed using ``MyDataFlow::flow(DataFlow::Node source, DataFlow::Node sink)``:
```

Reply:
This is the change I'm least sure about, since "instantiates the parameterized module" really is the thing that line of code is doing, whereas "passes the configuration" is an important but secondary detail.

@MathiasVP I believe you commented on this bit previously - do you have an opinion?

Reply:
Hm. I agree with @geoffw0. I don't think this change makes the sentence more clear. I'd much rather keep the phrasing as-is (if @felicitymay don't mind).

Reply:
Thanks. Resolving this comment.